include/pfn/expected.hpp

```diff
@@ -0,0 +1,1483 @@
+  {
+    if (not set_) {
+      e_ = ::std::forward<G>(s.error());
+    } else {
+      _reinit(e_, v_, ::std::forward<G>(s.error()));
+      set_ = false;
+    }
+    return *this;
+  }
+
+  template <class... Args>
+  constexpr T &emplace(Args &&...args) noexcept
+    requires(::std::is_nothrow_constructible_v<T, Args...>)
+  {
+    if (set_) {
+      ::std::destroy_at(::std::addressof(v_));
+    } else {
+      ::std::destroy_at(::std::addressof(e_));
+      set_ = true;
+    }
+    return *::std::construct_at(::std::addressof(v_), std::forward<Args>(args)...);
+  }
+
+  template <class U, class... Args>
+  constexpr T &emplace(::std::initializer_list<U> il, Args &&...args) noexcept
+    requires(::std::is_nothrow_constructible_v<T, ::std::initializer_list<U> &, Args...>)
+  {
+    if (set_) {
+      ::std::destroy_at(::std::addressof(v_));
+    } else {
+      ::std::destroy_at(::std::addressof(e_));
+      set_ = true;
+    }
+    return *::std::construct_at(::std::addressof(v_), il, std::forward<Args>(args)...);
+  }
+
+  // [expected.object.swap], swap
+  constexpr void
+  swap(expected &rhs) noexcept(::std::is_nothrow_move_constructible_v<T> && ::std::is_nothrow_swappable_v<T>
+                               && ::std::is_nothrow_move_constructible_v<E> && ::std::is_nothrow_swappable_v<E>)
+    requires(::std::is_swappable_v<T> && ::std::is_swappable_v<E> && ::std::is_move_constructible_v<T>
+             && ::std::is_move_constructible_v<E>
+             && (::std::is_nothrow_move_constructible_v<T> || ::std::is_nothrow_move_constructible_v<E>))
+  {
+    bool const lhset = has_value();
+    bool const rhset = rhs.has_value();
+    if (lhset == rhset) {
+      if (lhset) {
+        using ::std::swap;
+        swap(v_, rhs.v_);
+      } else {
+        using ::std::swap;
+        swap(e_, rhs.e_);
+      }
+    } else {
+      if (lhset) {
+        _swap(*this, rhs);
+      } else {
+        _swap(rhs, *this);
+      }
+    }
+  }
+
+  constexpr friend void swap(expected &x, expected &y) noexcept(noexcept(x.swap(y)))
+    requires requires { x.swap(y); }
+  {
+    x.swap(y);
+  }
+
+  // [expected.object.obs], observers
+  constexpr T const *operator->() const noexcept
+  {
+    ASSERT(set_);
+    return ::std::addressof(v_);
+  }
+  constexpr T *operator->() noexcept
+  {
+    ASSERT(set_);
+    return ::std::addressof(v_);
+  }
+  constexpr T const &operator*() const & noexcept { return *(this->operator->()); }
+  constexpr T &operator*() & noexcept { return *(this->operator->()); }
+  constexpr T const &&operator*() const && noexcept { return ::std::move(*(this->operator->())); }
+  constexpr T &&operator*() && noexcept { return ::std::move(*(this->operator->())); }
+  constexpr explicit operator bool() const noexcept { return set_; }
+  constexpr bool has_value() const noexcept { return set_; }
+  constexpr T const &value() const &
+  {
+    static_assert(::std::is_copy_constructible_v<E>);
```

Comment:
Do you prefer `static_assert` over requires clause for this and below?

Reply:
i think `requires` removes the function from the overload candidate list if it doesn’t meet the requirements, which doesn’t align with the spec

Reply:
@a1q123456 is right. The C++ standard prescribes here `Mandates`, see https://eel.is/c++draft/expected.object.obs , which in laymen terms means that "if not met, the program is ill-formed". https://eel.is/c++draft/description#structure.specifications-3.2 . Moving this condition check to `requires` would make this implementation non-conforming.